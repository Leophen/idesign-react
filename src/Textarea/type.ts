export interface TextareaProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 占位符
   * @default 请输入
   */
  placeholder?: string;
  /**
   * 文本框固定值（受控）
   */
  value?: string | number;
  /**
   * 文本框默认值（非受控）
   */
  defaultValue?: string | number;
  /**
   * 是否禁用文本框
   * @default false
   */
  disabled?: boolean;
  /**
   * 文本框是否只读
   * @default false
   */
  readonly?: boolean;
  /**
   * 文本框状态
   */
  status?: 'success' | 'warning' | 'error';
  /**
   * 文本框底部提示
   */
  tips?: string;
  /**
   * 用户最多可以输入的字符个数
   */
  maxLength?: number;
  /**
   * 自动聚焦
   * @default false
   */
  autofocus?: boolean;
  /**
   * 是否禁用右下角高度调整模块
   * @default false
   */
  noResize?: boolean;
  /**
   * 自适应高度
   * @default false
   */
  autoSize?: boolean;
  /**
   * 最小高度
   */
  minRows?: number;
  /**
   * 最大高度
   */
  maxRows?: number;
  /**
   * 输入内容变化时触发
   */
  onChange?: (value: string | number, context?: { e?: React.FormEvent<HTMLDivElement> }) => void;
  /**
   * 输入框聚焦时触发
   */
  onFocus?: (value: string | number, context?: { e?: React.FocusEvent<HTMLInputElement> }) => void;
  /**
   * 输入框失焦时触发
   */
  onBlur?: (value: string | number, context?: { e?: React.FocusEvent<HTMLInputElement> }) => void;
}
