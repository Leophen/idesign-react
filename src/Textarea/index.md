---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 4
---

# Textarea 多行文本框

用于字符输入的多行输入框。

## 基本用法

用于字符输入的多行输入框：

```tsx
import React, { useState } from 'react'
import { Textarea } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('这是默认值')

  const handleChange = (val, e) => {
    console.log(val, e)
    setValue(val)
  }

  return (
    <>
      <h4>无默认值</h4>
      <Textarea />
      <h4>有默认值</h4>
      <Textarea defaultValue={value} />
      <h4>有受控值</h4>
      <Textarea value={value} />
      <h4>一般用法</h4>
      <Textarea value={value} onChange={handleChange} />
    </>
  )
}

export default App
```

## 禁用状态

通过 `disabled` 属性指定是否禁用输入框。

```tsx
import React from 'react';
import { Textarea } from 'idesign-react';

const App = () => {
  return <Textarea disabled />;
};

export default App;
```

## 只读状态

通过 `readonly` 属性指定是否为只读状态。

```tsx
import React from 'react';
import { Textarea } from 'idesign-react';

const App = () => {
  return <Textarea readonly />;
};

export default App;
```

## 不同状态

通过 `status` 属性指定不同状态的输入框，通过 `tips` 属性指定 底部提示信息。

```tsx
import React from 'react';
import { Textarea } from 'idesign-react';

const App = () => {
  return (
    <>
      <Textarea placeholder="成功状态" status="success" />
      <br />
      <Textarea placeholder="警告状态" status="warning" />
      <br />
      <Textarea placeholder="错误状态" status="error" />
      <br />
      <Textarea placeholder="成功状态" status="success" tips="成功状态提示" />
      <Textarea placeholder="警告状态" status="warning" tips="警告状态提示" />
      <Textarea placeholder="错误状态" status="error" tips="错误状态提示" />
    </>
  );
};

export default App;
```

## 限制字数

通过 `maxLength` 属性指定最多可以输入的字符个数。

```tsx
import React, { useState } from 'react';
import { Textarea } from 'idesign-react';

const App = () => {
  const [value, setValue] = useState('iDesign');

  return (
    <>
      <Textarea maxLength={15} />
      <br />
      <Textarea
        value={value}
        maxLength={20}
        onChange={(value) => {
          console.log(value);
          setValue(value);
        }}
      />
    </>
  );
};

export default App;
```

## 禁用高度调整

通过 `noResize` 属性指定是否禁用右下角高度调整模块。

```tsx
import React from 'react';
import { Textarea } from 'idesign-react';

const App = () => {
  return <Textarea noResize />;
};

export default App;
```

## 限制高度

通过 `minRows` 和 `maxRows` 属性控制最小和最大高度。

```tsx
import React from 'react';
import { Textarea } from 'idesign-react';

const App = () => {
  return (
    <>
      <Textarea
        placeholder="限制最小高度，minRows={2}"
        minRows={2}
      />
      <br />
      <Textarea
        placeholder="限制最大高度，maxRows={6}"
        maxRows={6}
      />
      <br />
      <Textarea
        placeholder="限制最小和最大高度，minRows={2} maxRows={6}"
        minRows={2}
        maxRows={6}
      />
    </>
  );
};

export default App;
```

## 高度自适应

通过 `autoSize` 属性控制高度自适应。

```tsx
import React from 'react';
import { Textarea } from 'idesign-react';

const App = () => {
  return (
    <>
      <Textarea
        placeholder="高度自适应，autoSize = true"
        autoSize
      />
      <br />
      <Textarea
        placeholder="高度自适应 + 限制高度，minRows={2} maxRows={6}"
        autoSize
        minRows={2}
        maxRows={6}
      />
    </>
  );
};

export default App;
```

## 触发事件

一系列的事件触发操作。

```tsx
import React, { useState } from 'react';
import { Textarea } from 'idesign-react';

const App = () => {
  const [value, setValue] = useState('iDesign');

  return (
    <>
      <h3>onChange 输入时触发</h3>
      <Textarea
        onChange={(val, e) => {
          console.log('onChange', val, e);
        }}
      />
      <h3>onFocus 聚焦时触发</h3>
      <Textarea
        onFocus={(val, e) => {
          console.log('onFocus', val, e);
        }}
      />
      <h3>onBlur 失焦时触发</h3>
      <Textarea
        onBlur={(val, e) => {
          console.log('onBlur', val, e);
        }}
      />
    </>
  );
};

export default App;
```

<API />
