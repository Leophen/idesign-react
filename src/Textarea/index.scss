@mixin settingStatus($color) {
  &,
  &:hover,
  &:focus-within {
    border-color: $color;
  }
}

.i-textarea {
  position: relative;
  width: 100%;
}

.i-textarea__inner {
  display: flex;
  width: 100%;
  height: 56px;
  min-height: 56px;
  padding: 5px 8px;
  color: inherit;
  color: var(--i-font-2);
  font-size: 14px;
  font-family: auto;
  line-height: 22px;
  background-color: var(--i-bg);
  border: 1px solid var(--i-border);
  border-radius: 4px;
  outline: none;
  transition: all 0.2s cubic-bezier(0.38, 0, 0.24, 1), height 0s;
  resize: vertical;
  &::placeholder {
    color: var(--i-font-5);
    user-select: none;
  }
  &:hover {
    border-color: var(--i-border-hover);
  }
  &:focus {
    border-color: var(--i-primary);
  }
  &.i-textarea__inner-is-disabled {
    background-color: var(--i-bg-disabled);
    &,
    * {
      cursor: not-allowed;
    }
    &:hover {
      border-color: var(--i-border);
    }
  }
  &.i-textarea__inner-is-readonly {
    &,
    * {
      cursor: default;
    }
  }
  &.i-textarea__inner--status-success {
    @include settingStatus(var(--i-success));
  }
  &.i-textarea__inner--status-warning {
    @include settingStatus(var(--i-warning));
  }
  &.i-textarea__inner--status-error {
    @include settingStatus(var(--i-error));
  }
}

.i-textarea--limit {
  position: absolute;
  right: 0px;
  bottom: 8px;
  margin-left: 8px;
}

.i-textarea--limit-count {
  color: var(--i-font-5);
  font-size: 12px;
}

.i-textarea__tips {
  height: auto;
  min-height: 24px;
  color: var(--i-font-5);
  font-size: 12px;
  line-height: 20px;
  &.i-textarea__tips--status-success {
    color: var(--i-success);
  }
  &.i-textarea__tips--status-warning {
    color: var(--i-warning);
  }
  &.i-textarea__tips--status-error {
    color: var(--i-error);
  }
}
