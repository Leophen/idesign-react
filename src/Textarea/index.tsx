import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import useDefault from '../hooks/useDefault';
import { TextareaProps } from './type';

const Textarea: React.FC<TextareaProps> = (props) => {
  const {
    className,
    style,
    placeholder = "请输入",
    disabled = false,
    readonly = false,
    status,
    tips,
    maxLength,
    autofocus = false,
    autoSize = false,
    noResize = false,
    minRows,
    maxRows,
    value,
    defaultValue = '',
    onChange,
    onFocus,
    onBlur,
    ...restProps
  } = props;

  const [innerValue, setInnerValue] = useDefault(value, defaultValue, onChange);

  const [valueLength, setValueLength] = useState(innerValue?.toString().length || 0);

  const handleChange = (e: React.FormEvent<HTMLTextAreaElement>) => {
    e.persist();
    maxLength && setValueLength((e.target as HTMLTextAreaElement).value.length);
    setInnerValue((e.target as HTMLTextAreaElement).value, e as any)
    resize()
  };

  // 高度自适应
  useEffect(() => {
    autoSize ? resize() : ''
  }, [autoSize])
  const textareaRef = useRef(null)
  const resize = () => {
    if (autoSize && textareaRef.current) {
      // 如果字数减少，重设高度
      (textareaRef.current as HTMLElement).style.height = 'auto';
      // 如果高度不够，重新设置
      if ((textareaRef.current as HTMLElement).scrollHeight >= (textareaRef.current as HTMLElement).offsetHeight) {
        let val;
        (textareaRef.current as HTMLElement).scrollHeight > 32 ? val = (textareaRef.current as HTMLElement).scrollHeight - 10 : val = (textareaRef.current as HTMLElement).scrollHeight;
        (textareaRef.current as HTMLElement).style.height = val + 'px'
      }
    }
  }

  // 通用事件
  const handleEvent = (eventType: 'focus' | 'blur', e: any) => {
    e.persist();
    if (eventType === 'focus') {
      onFocus?.(e.target.value, e);
    }
    if (eventType === 'blur') {
      onBlur?.(e.target.value, e);
    }
  };

  return (
    <>
      <div
        className={classNames(
          'i-textarea',
          className
        )}
        style={{ ...style }}
        {...restProps}
      >
        <textarea
          className={classNames(
            'i-textarea__inner',
            disabled && 'i-textarea__inner-is-disabled',
            readonly && 'i-textarea__inner-is-readonly',
            status && `i-textarea__inner--status-${status}`,
          )}
          style={{
            minHeight: autoSize ? minRows && (22 * minRows) || 22 : minRows && (22 * minRows),
            maxHeight: maxRows && (22 * maxRows),
            resize: noResize ? 'none' : undefined
          }}
          placeholder={placeholder}
          disabled={disabled}
          readOnly={readonly}
          ref={textareaRef}
          rows={autoSize ? 1 : undefined}
          maxLength={maxLength}
          autoFocus={autofocus}
          value={innerValue}
          onChange={handleChange}
          onFocus={(e) => handleEvent('focus', e)}
          onBlur={(e) => handleEvent('blur', e)}
        ></textarea>
        {maxLength && (
          <div className="i-textarea--limit">
            <span className="i-textarea--limit-count">
              {valueLength < maxLength ? valueLength : maxLength} / {maxLength}
            </span>
          </div>
        )}
      </div>
      {tips && <div
        className={classNames(
          'i-textarea__tips',
          status && `i-textarea__tips--status-${status}`,
        )}
      >
        {tips}
      </div>}
    </>
  );
};

Textarea.displayName = 'Textarea';

export default Textarea;
