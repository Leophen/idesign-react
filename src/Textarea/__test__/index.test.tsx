import { render, fireEvent } from '@testing-library/react';
import Textarea from '../index';

describe('Textarea 组件测试', () => {
  const renderTextarea = (textarea) => {
    const { container } = render(textarea);
    return container.querySelector('.i-textarea__inner')
  }

  it('create', () => {
    const { asFragment } = render(<Textarea />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('placeholder onChange', () => {
    const changeFn = jest.fn();
    const { queryByPlaceholderText } = render(<Textarea placeholder='请输入内容' onChange={changeFn} />);
    expect(queryByPlaceholderText('请输入内容')).toBeInTheDocument();
    fireEvent.change(queryByPlaceholderText('请输入内容'), { target: { value: 'iDesign' } });
    expect(changeFn).toBeCalledTimes(1);
    expect(changeFn.mock.calls[0][0]).toBe('iDesign');
  });

  it('value', () => {
    expect(renderTextarea(<Textarea value='iDesign' />).value).toEqual('iDesign');
  });

  it('defaultValue', () => {
    expect(renderTextarea(<Textarea defaultValue='iDesign' />).value).toEqual('iDesign');
  });

  it('disabled', () => {
    const focusFn = jest.fn();
    const { container } = render(<Textarea disabled onFocus={focusFn} />);
    expect(container.firstChild.firstChild).toHaveClass('i-textarea__inner-is-disabled');
    fireEvent.click(container.firstChild);
    expect(focusFn).toBeCalledTimes(0);
  });

  it('readonly', () => {
    expect(renderTextarea(<Textarea readonly />)).toHaveClass('i-textarea__inner-is-readonly');
  });

  it('maxLength', () => {
    const { container } = render(<Textarea maxLength={15} />);
    expect(container.querySelector('.i-textarea--limit')).toBeInTheDocument();
  })

  it('status', () => {
    expect(renderTextarea(<Textarea status="success" />)).toHaveClass('i-textarea__inner--status-success');
  });

  it('status-tips', () => {
    render(<Textarea status="success" tips="成功状态提示" />);
    expect(document.querySelector('.i-textarea__tips')).toBeInTheDocument();
  });

  it('noResize', () => {
    expect(renderTextarea(<Textarea noResize />)).toHaveStyle('resize: none');
  });
});

