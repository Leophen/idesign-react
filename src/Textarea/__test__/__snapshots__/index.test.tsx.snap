// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-textarea"
  >
    <textarea
      class="i-textarea__inner"
      placeholder="请输入"
    />
  </div>
</DocumentFragment>
`;
